File: LC/Python/Hard/115.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        m = len(s)
        n = len(t)
        cache = {}
        
        def dfs(i, j):
            if j == n:
                return 1
            if i == m:
                return 0
            if (i, j) in cache:
                return cache[(i, j)]
            if m - i < n - j:
                return 0
            
            res = dfs(i + 1, j)
            if s[i] == t[j]:
                res += dfs(i + 1, j + 1)
            
            cache[(i, j)] = res
            
            return res
            
        
        return dfs(0 ,0)

File: LC/C++/Medium/115.cpp
class Solution {
    size_t m;
    size_t n;
    std::vector<std::vector<int>> cache_;
public:
    int dfs(string& s, string& t, size_t i, size_t j) {
        if (j == n) return 1;
        if (i == m) return 0;
        if (cache_[i][j] != -1) return cache_[i][j];
        if ((m - i) < (n - j)) return 0;

        int res = dfs(s, t, i + 1, j);
        if (s[i] == t[j]) res += dfs(s, t, i + 1, j + 1);

        return cache_[i][j] = res;
    }

    int numDistinct(string& s, string& t) {
        m = s.size();
        n = t.size();
        cache_.resize(m, std::vector(n, -1));

        return dfs(s, t, 0, 0);
    }
};